### PR TITLE
[RLlib] Unify GPU handling by RLlib (num_gpus; automate GPU testing).

### PR DESCRIPTION
**Pull request by sven1977**
This PR changes the default value of `num_gpus` in the Trainer's config to `None`.
Allowed values:

```
None: Use as many GPUs as are available (this used to be the default behavior anyways for TFPolicy, which never cared about e.g. 
num_gpus=0 and would still run on the GPU if one was available).
0: Do NOT use a GPU, even if one is available. TFPolicy and TorchPolicy used to violate this, causing some confusion amongst users who had force-set this to 0.
[some int]: Use this many GPUs. Error if less are available.
```

Also adds the option of setting an env variable (`RLLIB_FORCE_NUM_GPUS`) to some value to override `num_gpus` inside all RLlib's example scripts to use that many GPUs for upcoming daily GPU testing.
(TODO: Do the same for all test cases.)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/utils/test_utils.py [hunk 1]
```diff
@@ -16,6 +17,8 @@
 
 logger = logging.getLogger(__name__)
 
+FORCED_NUM_GPUS = int(os.environ.get("RAY_FORCE_NUM_GPUS", 0)) or None
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/test_utils.py:20]: +1

File: rllib/agents/trainer.py [hunk 2]
```diff
@@ -1078,6 +1055,18 @@ def make_callbacks():
 
     @staticmethod
     def _validate_config(config: PartialTrainerConfigDict):
+        # Resolve `num_gpus`=None (default): Set `num_gpus` to as many as we
```
Comment (sven1977) [inline rllib/agents/trainer.py:1058]: This is the crux of this PR: How do we figure out, how many GPUs are available if `num_gpus: None`?

File: rllib/policy/torch_policy.py [hunk 3]
```diff
@@ -104,9 +107,12 @@ def __init__(
         """
         self.framework = "torch"
         super().__init__(observation_space, action_space, config)
-        if torch.cuda.is_available() and ray.get_gpu_ids():
+        if ray.get_gpu_ids():
```
Comment (sven1977) [inline rllib/policy/torch_policy.py:110]: In TorchPolicy: Decide inside Policy constructor, whether to place the Model on the GPU.

File: rllib/examples/attention_net.py [hunk 4]
```diff
@@ -8,7 +8,8 @@
 from ray.rllib.examples.env.repeat_initial_obs_env import RepeatInitialObsEnv
 from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole
 from ray.rllib.utils.framework import try_import_tf
-from ray.rllib.utils.test_utils import check_learning_achieved
+from ray.rllib.utils.test_utils import check_learning_achieved, \
```
Comment (sven1977) [inline rllib/examples/attention_net.py:11]: Fix all example scripts and Model classes in `examples/model` to work with env var RLLIB_FORCE_NUM_GPUS=[1|0|None].

File: rllib/agents/trainer.py [hunk 5]
```diff
@@ -584,32 +586,6 @@ def setup(self, config: PartialTrainerConfigDict):
         self.config = Trainer.merge_trainer_configs(self._default_config,
                                                     config)
 
-        # Check and resolve DL framework settings.
```
Comment (sven1977) [inline rllib/agents/trainer.py:587]: Framework checks (installations vs config) moved to separate method.

File: rllib/agents/trainer.py [hunk 6]
```diff
@@ -81,7 +81,9 @@
     # Number of GPUs to allocate to the trainer process. Note that not all
     # algorithms can take advantage of trainer GPUs. This can be fractional
     # (e.g., 0.3 GPUs).
-    "num_gpus": 0,
+    # None (default): Use all available GPUs.
+    # 0: Will not use any GPUs, even if one is available.
+    "num_gpus": None,
```
Comment (sven1977) [inline rllib/agents/trainer.py:86]: New default for num_gpus: None -> Figure out GPU placement based on what's available.
Also: num_gpus=0 means: Do NOT use any GPUs.

**Review by ericl — CHANGES_REQUESTED** [review]
This is a pretty major change. Why not instead simply raise a warning instead if a GPU is present but it is set to zero?

**Comment by ericl** [comment]
Couple things:
(1) Can you split this into two PRs -- one for the GPU interface change, and the other for the GPU testing? We shouldn't be mixing major API changes with other big code changes.
(2) I was thinking the zero GPU case could be simply a warning in Trainer -- to let the user they had left num_gpus=0 if they didn't mean it. That would be just a couple lines and would not require any changes to the algorithms.

**Pull request closed without merging**